chalk-ir/src/fold.rs

```diff
@@ -376,3 +314,70 @@ impl<TF: TypeFamily, TTF: TargetTypeFamily<TF>> Fold<TF, TTF> for Lifetime<TF> {
         folder.fold_lifetime(self, binders)
     }
 }
+
+impl<TF, TTF> SuperFold<TF, TTF> for Lifetime<TF>
+where
+    TF: TypeFamily,
+    TTF: TargetTypeFamily<TF>,
+{
+    fn super_fold_with(
+        &self,
+        folder: &mut dyn Folder<TF, TTF>,
+        binders: usize,
+    ) -> Fallible<Lifetime<TTF>> {
+        match self.data() {
+            LifetimeData::BoundVar(depth) => {
+                if *depth >= binders {
+                    folder.fold_free_var_lifetime(depth - binders, binders)
+                } else {
+                    Ok(LifetimeData::<TTF>::BoundVar(*depth).intern())
+                }
+            }
+            LifetimeData::InferenceVar(var) => folder.fold_inference_lifetime(*var, binders),
+            LifetimeData::Placeholder(universe) => {
+                folder.fold_free_placeholder_lifetime(*universe, binders)
+            }
+            LifetimeData::Phantom(..) => unreachable!(),
+        }
+    }
+}
+
+/// Folding a goal invokes the `fold_goal` callback (which will, by
+/// default, invoke super-fold).
+impl<TF: TypeFamily, TTF: TargetTypeFamily<TF>> Fold<TF, TTF> for Goal<TF> {
```

Comment:
Why remove this from `boring_impls`?

Reply:
It's not boring. It invokes a callback on folder rather than just recursively reproducing its structure.

Reply:
Maybe this naming scheme isn't obvious to everyone else though. :)

Reply:
Yeah, my thought for "boring" is either boilerplate-like code or code that is extremely simple.